src/clj/fluree/db/flake/reasoner.cljc

```diff
@@ -163,28 +165,45 @@
     (log/debug "Reasoner - source OWL rules: " graph)
     (owl-datalog/owl->datalog inserts graph)))
 
+(defn rules-from-dbs
+  [methods inserts dbs]
+  (let [rules (atom [])]
```

Comment:
I don't think you need an atom here, you can accumulate the data you want in the loops - here's a rough example:
```clojure
(go-try
  (loop [[method & r-methods] methods
         rules []]
    (if method
      (loop [[db & r-dbs] dbs
             rules* rules]
        (if db
          (let [db-rules (<? (resolve/rules-from-db db method))]
            (recur r-dbs (into rules* (rules-from-graph method inserts db-rules))))
          (into rules rules*)))
      rules)))
```

Reply:
This feels more complicated than the code I had before wrapping a `for` loop in `doall` to remove the laziness. I think that also reads better than this. Is there a reason that isn't ideal?

Reply:
Can you elaborate on what feels more complicated? . This imperative-style mutation-with-a-loop is pretty unidiomatic in Clojure. Mutation is inherently complex and if we're not doing it for performance reasons (and I believe in this case it's actually slower, though probably not critically so) it's good practice to avoid it. 

I also think it's simpler  - it's less code, the structure is pretty clear, and it should be plenty fast.

Reply:
I'm thinking the for-loop would look something like this:

Reply:
Ok, that explains some of the issues I've been running into trying to get it to work. All this async behavior seems really finicky and unintuitive on when it works and when it doesn't.

Reply:
I find it much easier to read recursive functions if the `recur` doesn't contain multi-line expressions. Can you restructure this along the lines of what I sketched out earlier?